### DISCUSSION TITLE
Managing dependencies between configuration and source repositories

### DISCUSSION THREAD

**Discussion by onionringois** [argoproj/argo-cd#17797] [Q&A]
**To give you some background,**
we work in an architecture that separates the code into one repository and the helm chart (configuration) into another repository in accordance with the GitOps methodology.

Right now, as part of our CD process, there is a stage in gitlab ci that updates the helm chart value with the, newly built image, and applies all the changes made since the last deploy. 
This means that any changes we make to the helm chart won't be visible to k8s cluster until we deploy from pipeline that was created from the source repository.
The good thing about the current scenario is that I have control over the changes that are made to our deploy environments, so I can avoid situations where a "version" of the config map is not aligned to the suitable image version thus prevent crashloop situations due to lack / unexpected change of configuration.

One issue that comes up when using Argo CD with autosync is that I can't plan a code change and a configuration update at the same time since if I change the values, Argo will modify the configmap and accordingly will also rollout the deployment (we have an annotation in the deployment whose value is the sha of the configmap so that the deployment will rollout with every change of the configmap through the helm) with the old version of the image.

**In summary, there are situations in which the source and the configuration are dependent on one another.**

One solution to switch from autosync to manual sync. 
By doing so, we will be able to precisely control when the configmap change is applied and even trigger a sync when the CI process is complete after updating the image.
In this arrangement, there won't be any inconsistency between the "version" of the configmap and image version.

The problem with this solution is that it significantly reduces the benefits of our use of Argo and pretty much leaves it in the same state where we currently work only with Argo's UI and some other nice features.
It is best for us to continue on our current work path in this circumstance.

I would appreciate any suggestions for a fix or a general improvement.
Perhaps we don't have a clear perspective on all of the gitops methodology.

**As an aside**, we are unable to use Argo rollouts at the moment.

**Thank you!!**

**Comment by crenshaw-dev** [comment]
This sounds like a more specific version of the "database schema update" change. The general solution is to write your app code in a way that is compatible with both states of the external system (in your case, with the old ConfigMap state and with the new ConfigMap state). The alternative is to rely on some other synchronization system, whether that's a human manually clicking the Argo CD "Sync" button, staging image changes in a PR where related changes can be appended, or something else. I'm not sure that a general standard has emerged for all such cases.
